[clang-tidy] Support find for string-like classes in readability-container-contains

nicovank opened the pull request:
Fix #109327. Small parts taken from #110351.

Removed the type checking between `contains` and `count`/`find` arguments for simplicity. Because of overloads, this type-checking is tricky. The same strategy is used in `modernize-use-starts-ends-with`.

Ended up copying over the FixIt logic from `modernize-use-starts-ends-with`, I think it's the simplest. Open to other suggestions.

Since this may introduce false positives in (bad readability) edge cases, I tested this change on several major C++ projects, and didn't find any new build or test failures introduced by this change (some errors due to C++23 upgrade, unrelated). Tested projects:
 1. kitware/cmake
 2. dolphin-emu/dolphin
 3. blender/blender
 4. opencv/opencv

Co-authored-by: dl8sd11 <gcchen@google.com>

EugeneZelenko commented on clang-tools-extra/clang-tidy/readability/ContainerContainsCheck.cpp:
```diff
@@ -114,29 +112,39 @@ void ContainerContainsCheck::check(const MatchFinder::MatchResult &Result) {
-  SourceLocation ComparisonEnd = Comparison->getSourceRange().getEnd();
-  SourceLocation CallBegin = Call->getSourceRange().getBegin();
-  SourceLocation CallEnd = Call->getSourceRange().getEnd();
+  const auto SearchExprText = Lexer::getSourceText(
```
Please do not use `auto` unless type explicitly stated in same statement or iterator.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -230,6 +230,10 @@ Changes in existing checks
   <clang-tidy/checks/portability/template-virtual-member-function>` check to
   avoid false positives on pure virtual member functions.
 
+- Fixed false negatives on string and string-like types in
```
Please follow format of other similar statements.

vbvictor commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -263,6 +263,11 @@ Changes in existing checks
 
+- Improved :doc:`readability-container-contains
+  <clang-tidy/checks/readability/container-contains>` to support string
+  comparisons to ``npos``. Internal changes may cause new rare false positives
+  in non-standard containers.
```
I kinda don't like that we deliberately allow false-positives, but if it works in `use-starts-ends-with`, let it be it

nicovank commented:
Test failures seem unrelated after rebase on main. I'll try again a little bit later.

nicovank committed:
[clang-tidy] Support find for string-like classes in readability-container-contains

Fix #109327. Small parts taken from #110351.

Removed the type checking between `contains` and `count`/`find` arguments for simplicity. Because of overloads, this type-checking is tricky. The same strategy is used in `modernize-use-starts-ends-with`.

Ended up copying over the FixIt logic from `modernize-use-starts-ends-with`, I think it's the simplest. Open to other suggestions.

Since this may introduce false positives in (bad readability) edge cases, I tested this change on several major C++ projects, and didn't find any new build or test failures introduced by this change (some errors due to C++23 upgrade, unrelated). Tested projects:
 1. kitware/cmake
 2. dolphin-emu/dolphin
 3. blender/blender
 4. opencv/opencv

Co-authored-by: dl8sd11 <gcchen@google.com>
